core/modules/link/link.module

```diff
@@ -420,6 +439,20 @@ function _link_validate(&$item, $delta, $field, $entity, $instance, $langcode, &
         'error_element' => array('url' => FALSE, 'title' => TRUE),
       );
     }
+    // Validate the URL type.
+    if ($instance['settings']['type'] != LINK_BOTH) {
+      $url_type = link_validate_url(trim($item['url']));
+      if (($instance['settings']['type'] == LINK_INTERNAL && $url_type === LINK_EXTERNAL) || ($instance['settings']['type'] == LINK_EXTERNAL && in_array($url_type, array(LINK_INTERNAL, LINK_FRONT), TRUE))) {
+        $errors[$field['field_name']][$langcode][$delta][] = array(
+          'error' => 'link_required',
+          'message' => t('Only %type URLs are allowed in %field.', array(
```

Comment:
This error message has the same problem as above where "internal" and "external" needs to be translatable. This would probably be best as two strings based on the link type.